### PR TITLE
Portability fixes

### LINKED ISSUE #2: configuration error on macos, because of missing (needless) libbsd
```
> aclocal
...
> autoreconf -i
...
> autoreconf
...
./configure --prefix=/test --with-openssl=/opt/libressl
...
checking whether compiling and linking against OpenSSL works... yes
checking for pkg-config... /usr/local/bin/pkg-config
checking for bsd/bsd.h in /usr/local/bsd... no
checking for bsd/bsd.h in /usr/lib/bsd... no
checking for bsd/bsd.h in /usr/bsd... no
checking for bsd/bsd.h in /usr/pkg... no
checking for bsd/bsd.h in /usr/local... no
checking for bsd/bsd.h in /usr... no
checking whether compiling and linking against libbsd works... no
```
```
>uname -a
Darwin alice 17.7.0 Darwin Kernel Version 17.7.0: Sun Jun  2 20:31:42 PDT 2019; root:xnu-4570.71.46~1/RELEASE_X86_64 x86_64
```

### PR DESCRIPTION
**Pull request by NattyNarwhal**
Completely replaces libbsd dependency with internal compat routines, and adds some more for it to work on Solaris (not fully tested) and AIX (tested), and tested on glibc distros too. The only library dependency now is OpenSSL. The compat routines were sourced from tmux with copyright headers intact.

Some source code is also slightly adjusted to fix compatibility with the Solaris compiler, as it is a bit stricter than gcc on warnings.

Nitpick: at least on AIX, should be built with `-D_LINUX_SOURCE_COMPAT`, but haven't padded this yet.

Would probably fix #2, but I don't have a Mac to verify.

### CODE REVIEW

File: compat.h [hunk 1]
```diff
@@ -1,6 +1,13 @@
 #ifndef COMPAT_H_fogjseogje4
 #define COMPAT_H_fogjseogje4 1
 
+#include <sys/types.h>
+#include <stdarg.h>
+
+#include "config.h"
+
+/* HACK: OpenSSL symbols pollute parser */
```
Comment (graywolf) [inline compat.h]: Could you please explain what you mean by this? I do not get it.

**Comment by graywolf** [comment]
Thank you for all the work you put into this! I'm working on merging this and will take deeper look tomorrow. So far I have just the two questions:

Removing the dependency on libbsd: `AIX/PASE can't use libbsd as it's not the fd.o one, skip it` sorry I do not understand what you mean by this. Libbsd just is not present on those platforms? Or what is the exact issue prevent it from being used?

The one question about `ACMECLIENT_PARSER`.

Thanks

**Comment by NattyNarwhal** [comment]
A libbsd is present on those systems but it is *not* the Freedesktop.org one. It's basically a 4.xBSD compat library, and it causes hard to debug errors if you try to use it.

**Comment by NattyNarwhal** [comment]
ACMECLIENT_PARSER works around an issue where `RSA` typedef in OpenSSL headers clobbers an enum value in parse.y.

**Comment by graywolf** [comment]
Ok I've went over the PR and made few changes. Could you try branch `improve-portability` at `https://git.sr.ht/~graywolf/acme-client-portable` and let me know if it still works for you?

I've tested it on debian and it works for me, but I would appreciate a check I did not break it on your platform(s).

**Comment by NattyNarwhal** [comment]
still builds and runs on AIX. can test other platforms later.

**Comment by NattyNarwhal** [comment]
Solaris with Studio/native non-GNU binutils is broken. It supports LTO, but only in limited scenarios, which tricks configure into adding it.
```
gmake  all-am
gmake[1]: Entering directory '/home/calvin/src/acme-client-portable'
cc -I/usr/include  -g -D_GNU_SOURCE=1 -Wall -Wextra -Wformat -Wmissing-declarations -Wmissing-prototypes -Wpointer-arith -Wshadow -Wsign-compare -Wstrict-prototypes -Wundef -Wunused -Wvla -flto -L/usr/lib  -o acme-client acme_client-acctproc.o acme_client-base64.o acme_client-certproc.o acme_client-chngproc.o acme_client-compat.o acme_client-dbg.o acme_client-dnsproc.o acme_client-fileproc.o acme_client-http.o acme_client-jsmn.o acme_client-json.o acme_client-key.o acme_client-keyproc.o acme_client-main.o acme_client-netproc.o acme_client-parse.o acme_client-revokeproc.o acme_client-util.o -lssl -lcrypto
cc: Warning: Option -flto passed to ld, if ld is invoked, ignored otherwise
ld: fatal: option '-flto' is incompatible with building a dynamic executable
gmake[1]: *** [Makefile:514: acme-client] Error 2
gmake[1]: Leaving directory '/home/calvin/src/acme-client-portable'
gmake: *** [Makefile:417: all] Error 2
```

**Comment by NattyNarwhal** [comment]
(FWIW, Solaris with GCC chokes on the system's native yacc includes on -Werror...)

**Comment by graywolf** [comment]
I've just disabled -flto on solaris. I think we are good to merge or did I miss anything?

**Comment by NattyNarwhal** [comment]
Should be equivalent now. I would also recommend adding this to handle AIX/PASE's malloc deficiency:

```
diff --git a/configure.ac b/configure.ac
index a62153e..d409855 100644
--- a/configure.ac
+++ b/configure.ac
@@ -62,6 +62,10 @@ case "$host" in
 	# flto seems broken on solaris, so do not use it:
 	# https://github.com/graywolf/acme-client-portable/pull/5#issuecomment-599510762
 	;;
+*aix*|*os400*)
+	# malloc is different, workaround
+	CPPFLAGS="$CPPFLAGS -D_LINUX_SOURCE_COMPAT"
+	;;
 *)
 	AX_ADD_CFLAG([-flto])
 	AX_ADD_LDFLAG([-flto])
```

**Comment by NattyNarwhal** [comment]
(Err, that whole sequence could be better placed instead of in the LTO block. But the idea is the same)

**Comment by graywolf** [comment]
Merged into master and new version (0.3.0) released which includes this change. Tarball is https://data.wolfsden.cz/sources/acme-client-0.3.0.tar.xz so feel free to try it from it if you feel like it.

Either way, thank you very much for contributing this fix/improvement, it's always nice to support more platforms :)

**Pull request closed without merging**